File: backend/prism/server/object_manager.py
"""Core Prism object manager for server-side state management."""

import asyncio
from collections.abc import Awaitable, Callable
from typing import Any

from prism.core.delta import DeltaComputer
from prism.core.protocol import (
    ClientMessage,
    DeltaMessage,
    ErrorMessage,
    FullObjectMessage,
    ServerMessage,
    SubscribeMessage,
    SyncMessage,
    UnsubscribeMessage,
    UpdateFilterMessage,
)
from prism.core.types import (
    ClientState,
    Delta,
    PrismObject,
    Subscription,
)
from prism.filters.base import FilterRegistry
from prism.server.cache import LRUCache
from prism.storage.base import StorageAdapter


class PrismObjectManager:
    """Manages object state, subscriptions, and synchronization for clients."""

    def __init__(
        self,
        storage: StorageAdapter,
        filters: FilterRegistry,
        version_cache_size: int = 10000,
        delta_cache_size: int = 5000,
        filter_cache_size: int = 5000,
    ):
        """Initialize object manager.

        Args:
            storage: Storage adapter for persisting objects
            filters: Filter registry
            version_cache_size: Size of version cache
            delta_cache_size: Size of delta cache
            filter_cache_size: Size of filter cache
        """
        self.storage = storage
        self.filters = filters
        self.clients: dict[str, ClientState] = {}
        self.version_cache: LRUCache[PrismObject] = LRUCache(version_cache_size)
        self.delta_cache: LRUCache[Delta] = LRUCache(delta_cache_size)
        self.filter_cache: LRUCache[PrismObject] = LRUCache(filter_cache_size)
        self.delta_computer = DeltaComputer()

        # Callbacks for sending messages to each client (per-client callback map)
        self.send_callbacks: dict[str, Callable[[ServerMessage], Awaitable[None]]] = {}

    def register_client(
        self, client_id: str, callback: Callable[[ServerMessage], Awaitable[None]]
    ) -> None:
        """Register a client's send callback.

        Args:
            client_id: Client identifier
            callback: Async function(message) for sending to this client
        """
        self.send_callbacks[client_id] = callback

    def get_client_state(self, client_id: str) -> ClientState:
        """Get or create client state.

        Args:
            client_id: Client identifier

        Returns:
            ClientState for this client
        """
        if client_id not in self.clients:
            self.clients[client_id] = ClientState()
        return self.clients[client_id]

    def remove_client(self, client_id: str) -> None:
        """Remove client state when disconnected.

        Args:
            client_id: Client identifier
        """
        self.clients.pop(client_id, None)
        self.send_callbacks.pop(client_id, None)

    async def handle_message(self, client_id: str, message: ClientMessage) -> None:
        """Handle incoming client message.

        Args:
            client_id: Client identifier
            message: The client message to handle
        """
        try:
            if isinstance(message, SubscribeMessage):
                await self.handle_subscribe(client_id, message)
            elif isinstance(message, UnsubscribeMessage):
                await self.handle_unsubscribe(client_id, message)
            elif isinstance(message, SyncMessage):
                await self.handle_sync(client_id, message)
            elif isinstance(message, UpdateFilterMessage):
                await self.handle_update_filter(client_id, message)
        except Exception as e:
            await self.send_error(client_id, "INTERNAL_ERROR", str(e))

    async def handle_subscribe(self, client_id: str, msg: SubscribeMessage) -> None:
        """Handle subscription request.

        Args:
            client_id: Client identifier
            msg: Subscribe message
        """
        # TODO: Add logging when logger is available
        # print(f"[SUBSCRIBE] Client {client_id} subscribing to {msg.object_id} with filter {msg.filter_type}")
        client = self.get_client_state(client_id)

        # Get current object
        current_obj = await self.storage.get_current(msg.object_id)
        if current_obj is None:
            await self.send_error(
                client_id, "OBJECT_NOT_FOUND", f"Object {msg.object_id} not found"
            )
            return

        # Apply filter
        filter_type = msg.filter_type or "default"
        filtered_obj = await self._apply_filter_cached(current_obj, filter_type, msg.filter_params)

        # Smart sync based on client state
        await self._smart_sync(client_id, msg.object_id, filtered_obj)

        # Register subscription if not temporary
        if not msg.temporary:
            subscription = Subscription(
                object_id=msg.object_id,
                filter_type=filter_type,
                filter_params=msg.filter_params,
                current_version=filtered_obj.version,
                temporary=False,
            )
            client.subscriptions[msg.object_id] = subscription
            client.update_version(msg.object_id, filtered_obj.version)

    async def handle_unsubscribe(self, client_id: str, msg: UnsubscribeMessage) -> None:
        """Handle unsubscribe request.

        Args:
            client_id: Client identifier
            msg: Unsubscribe message
        """
        client = self.get_client_state(client_id)
        client.subscriptions.pop(msg.object_id, None)

    async def handle_sync(self, client_id: str, msg: SyncMessage) -> None:
        """Handle sync request after reconnection.

        Args:
            client_id: Client identifier
            msg: Sync message with client's current state
        """
        client = self.get_client_state(client_id)

        for state in msg.states:
            current_obj = await self.storage.get_current(state.id)
            if current_obj is None:
                continue

            # Apply filter
            filtered_obj = await self._apply_filter_cached(current_obj, state.filter_type)

            # Smart sync
            await self._smart_sync(client_id, state.id, filtered_obj, state.version)

            # Re-register subscription (critical for receiving future updates)
            subscription = Subscription(
                object_id=state.id,
                filter_type=state.filter_type,
                current_version=filtered_obj.version,
                temporary=False,
            )
            client.subscriptions[state.id] = subscription
            client.update_version(state.id, filtered_obj.version)

    async def handle_update_filter(self, client_id: str, msg: UpdateFilterMessage) -> None:
        """Handle filter update for existing subscription.

        Args:
            client_id: Client identifier
            msg: Update filter message
        """
        client = self.get_client_state(client_id)

        if msg.object_id not in client.subscriptions:
            await self.send_error(
                client_id, "NOT_SUBSCRIBED", f"Not subscribed to {msg.object_id}"
            )
            return

        # Update subscription filter
        subscription = client.subscriptions[msg.object_id]
        subscription.filter_type = msg.filter_type
        subscription.filter_params = msg.filter_params

        # Re-sync with new filter
        current_obj = await self.storage.get_current(msg.object_id)
        if current_obj:
            filtered_obj = await self._apply_filter_cached(current_obj, msg.filter_type, msg.filter_params)
            # Always send full object when filter changes, even if version matches
            await self._send_full_object(client_id, filtered_obj)
            client.update_version(msg.object_id, filtered_obj.version)

    async def notify_object_updated(self, obj: PrismObject) -> None:
        """Notify all subscribed clients when an object is updated.

        Args:
            obj: The updated object
        """
        # Update caches
        self.version_cache.put(f"{obj.id}:{obj.version}", obj)

        # Notify all subscribed clients
        tasks = []
        for client_id, client_state in self.clients.items():
            if obj.id in client_state.subscriptions:
                subscription = client_state.subscriptions[obj.id]
                task = self._send_update(client_id, obj, subscription)
                tasks.append(task)

        if tasks:
            await asyncio.gather(*tasks, return_exceptions=True)

    async def _send_update(
        self, client_id: str, obj: PrismObject, subscription: Subscription
    ) -> None:
        """Send update to a specific client.

        Args:
            client_id: Client identifier
            obj: Updated object
            subscription: Client's subscription
        """
        # Apply filter
        filtered_obj = await self._apply_filter_cached(obj, subscription.filter_type, subscription.filter_params)

        # Smart sync
        await self._smart_sync(client_id, obj.id, filtered_obj)

    async def _smart_sync(
        self,
        client_id: str,
        object_id: str,
        current_obj: PrismObject,
        known_version: int | None = None,
    ) -> None:
        """Intelligently sync object with client based on their current state.

        Args:
            client_id: Client identifier
            object_id: Object ID
            current_obj: Current filtered object
            known_version: Version client has (if known)
        """
        client = self.get_client_state(client_id)

        if known_version is None:
            known_version = client.get_version(object_id)

        if known_version is None:
            # Client doesn't have object - send full
            await self._send_full_object(client_id, current_obj)
        elif known_version < current_obj.version:
            # Client has older version - try delta
            await self._send_delta_or_full(client_id, object_id, known_version, current_obj)
        # else: client has current version, no update needed

        # Update client's known version
        client.update_version(object_id, current_obj.version)

    async def _send_delta_or_full(
        self,
        client_id: str,
        object_id: str,
        from_version: int,
        to_obj: PrismObject,
    ) -> None:
        """Send delta if efficient, otherwise full object.

        Args:
            client_id: Client identifier
            object_id: Object ID
            from_version: Client's current version
            to_obj: Target object
        """
        # Try to compute delta
        delta = await self._compute_delta_cached(object_id, from_version, to_obj.version)

        if delta and self.delta_computer.is_delta_efficient(delta, to_obj):
            # Delta is efficient - send it
            await self._send_delta(client_id, delta)
        else:
            # Delta too large or unavailable - send full object
            await self._send_full_object(client_id, to_obj)

    async def _compute_delta_cached(
        self, object_id: str, from_version: int, to_version: int
    ) -> Delta | None:
        """Compute delta with caching.

        Args:
            object_id: Object ID
            from_version: Starting version
            to_version: Target version

        Returns:
            Delta or None if cannot be computed
        """
        cache_key = f"{object_id}:{from_version}:{to_version}"
        cached = self.delta_cache.get(cache_key)
        if cached:
            return cached

        # Get both versions
        from_obj = await self._get_version_cached(object_id, from_version)
        to_obj = await self._get_version_cached(object_id, to_version)

        if from_obj is None or to_obj is None:
            return None

        # Compute delta
        delta = self.delta_computer.compute_delta(from_obj, to_obj)
        self.delta_cache.put(cache_key, delta)
        return delta

    async def _get_version_cached(self, object_id: str, version: int) -> PrismObject | None:
        """Get object version with caching.

        Args:
            object_id: Object ID
            version: Version number

        Returns:
            PrismObject or None
        """
        cache_key = f"{object_id}:{version}"
        cached = self.version_cache.get(cache_key)
        if cached:
            return cached

        obj = await self.storage.get_version(object_id, version)
        if obj:
            self.version_cache.put(cache_key, obj)
        return obj

    async def _apply_filter_cached(
        self, obj: PrismObject, filter_type: str, filter_params: dict[str, Any] | None = None
    ) -> PrismObject:
        """Apply filter with caching.

        Args:
            obj: Object to filter
            filter_type: Filter name
            filter_params: Optional filter parameters

        Returns:
            Filtered object
        """
        if filter_type == "default":
            return obj

        # Include params in cache key if present
        cache_key = f"{obj.id}:{obj.version}:{filter_type}"
        if filter_params:
            # Simple hash of params for cache key
            import json
            params_str = json.dumps(filter_params, sort_keys=True)
            cache_key = f"{cache_key}:{params_str}"

        cached = self.filter_cache.get(cache_key)
        if cached:
            return cached

        filtered = self.filters.apply(obj, filter_type, filter_params)
        self.filter_cache.put(cache_key, filtered)
        return filtered

    async def _send_full_object(self, client_id: str, obj: PrismObject) -> None:
        """Send full object to client.

        Args:
            client_id: Client identifier
            obj: Object to send
        """
        message = FullObjectMessage(
            id=obj.id, version=obj.version, data=obj.data, filtered=True
        )
        await self._send_message(client_id, message)

    async def _send_delta(self, client_id: str, delta: Delta) -> None:
        """Send delta to client.

        Args:
            client_id: Client identifier
            delta: Delta to send
        """
        message = DeltaMessage(
            id=delta.object_id,
            from_version=delta.from_version,
            to_version=delta.to_version,
            patches=delta.patches,
        )
        await self._send_message(client_id, message)

    async def send_error(
        self, client_id: str, code: str, message: str, **kwargs: Any
    ) -> None:
        """Send error message to client.

        Args:
            client_id: Client identifier
            code: Error code
            message: Error message
            **kwargs: Additional error fields
        """
        error_msg = ErrorMessage(code=code, message=message, **kwargs)
        await self._send_message(client_id, error_msg)

    async def _send_message(self, client_id: str, message: ServerMessage) -> None:
        """Send message to client via callback.

        Args:
            client_id: Client identifier
            message: Message to send
        """
        callback = self.send_callbacks.get(client_id)
        if callback:
            await callback(message)

File: implementations/scala/core/src/main/scala/prism/server/ObjectManager.scala
package prism.server

import cats.effect.{IO, Ref}
import cats.syntax.all._
import prism.cache.LRUCache
import prism.core.DeltaComputer
import prism.core.Types._
import prism.core.Protocol._
import prism.filters.FilterRegistry
import prism.storage.StorageAdapter

/** Central manager for object storage, subscriptions, and notifications.
  *
  * The ObjectManager is the core of the Prism server, managing:
  *   - Object storage and versioning
  *   - Per-client subscription state
  *   - Filter application
  *   - Delta computation and caching
  *   - Notification of subscribers on updates
  *
  * Thread-safe implementation using cats-effect IO and Ref.
  *
  * @param storage
  *   Storage adapter for persisting objects
  * @param filterRegistry
  *   Registry of available filters
  * @param clientStates
  *   Mapping of client IDs to their state
  * @param sendCallbacks
  *   Mapping of client IDs to their send message callbacks
  * @param versionCache
  *   Cache for object versions
  * @param deltaCache
  *   Cache for computed deltas
  * @param filterCache
  *   Cache for filtered objects
  */
class ObjectManager private (
    storage: StorageAdapter,
    filterRegistry: FilterRegistry,
    clientStates: Ref[IO, Map[String, ClientState]],
    sendCallbacks: Ref[IO, Map[String, ServerMessage => IO[Unit]]],
    versionCache: LRUCache[String, PrismObject],
    deltaCache: LRUCache[(String, Int, Int, Option[String]), Delta],
    filterCache: LRUCache[(String, Int, String, Option[ujson.Value]), PrismObject]
) {

  /** Register a client's send callback for receiving server messages.
    *
    * @param clientId
    *   Client identifier
    * @param callback
    *   Function to send ServerMessage to this client
    */
  def registerClient(clientId: String, callback: ServerMessage => IO[Unit]): IO[Unit] = {
    sendCallbacks.update(_ + (clientId -> callback))
  }

  /** Get or create client state for a given client ID.
    *
    * @param clientId
    *   Unique client identifier
    * @return
    *   ClientState for this client
    */
  def getClientState(clientId: String): IO[ClientState] = {
    clientStates.modify { states =>
      states.get(clientId) match {
        case Some(state) => (states, state)
        case None =>
          val newState = new ClientState()
          (states + (clientId -> newState), newState)
      }
    }
  }

  /** Remove client state when a client disconnects.
    *
    * @param clientId
    *   Client identifier to remove
    */
  def removeClientState(clientId: String): IO[Unit] = {
    for {
      _ <- clientStates.update(_ - clientId)
      _ <- sendCallbacks.update(_ - clientId)
    } yield ()
  }

  /** Subscribe a client to an object.
    *
    * @param clientId
    *   Client identifier
    * @param objectId
    *   Object to subscribe to
    * @param filterType
    *   Filter to apply (default: "default")
    * @param filterParams
    *   Optional filter parameters
    * @param ongoing
    *   Whether this is an ongoing subscription (vs temporary)
    * @return
    *   The current object (filtered if requested)
    */
  def subscribe(
      clientId: String,
      objectId: String,
      filterType: String = "default",
      filterParams: Option[ujson.Value] = None,
      ongoing: Boolean = false
  ): IO[Option[PrismObject]] = {
    for {
      clientState <- getClientState(clientId)
      currentVersion <- storage.getCurrent(objectId).map(_.map(_.version))
      _ = currentVersion.foreach { version =>
        val subscription = Subscription(
          objectId = objectId,
          currentVersion = version,
          filterType = filterType,
          filterParams = filterParams,
          temporary = !ongoing
        )
        clientState.addSubscription(subscription)
      }
      obj <- storage.getCurrent(objectId)
      filtered <- obj match {
        case Some(o) => applyFilter(o, if (filterType == "default") None else Some(filterType), filterParams).map(Some(_))
        case None    => IO.pure(None)
      }
    } yield filtered
  }

  /** Unsubscribe a client from an object.
    *
    * @param clientId
    *   Client identifier
    * @param objectId
    *   Object to unsubscribe from
    */
  def unsubscribe(clientId: String, objectId: String): IO[Unit] = {
    for {
      clientState <- getClientState(clientId)
      _ = clientState.removeSubscription(objectId)
    } yield ()
  }

  /** Update the filter for a subscription.
    *
    * @param clientId
    *   Client identifier
    * @param objectId
    *   Object ID
    * @param filterType
    *   New filter type
    * @param filterParams
    *   New filter parameters
    */
  def updateFilter(
      clientId: String,
      objectId: String,
      filterType: String,
      filterParams: Option[ujson.Value]
  ): IO[Unit] = {
    for {
      clientState <- getClientState(clientId)
      sub <- IO(clientState.getSubscription(objectId))
      _ <- sub match {
        case Some(s) =>
          val updated = s.copy(filterType = filterType, filterParams = filterParams)
          IO(clientState.addSubscription(updated))
        case None =>
          IO.unit
      }
    } yield ()
  }

  /** Save an object and notify all subscribers.
    *
    * @param obj
    *   Object to save
    * @return
    *   List of client IDs that were notified
    */
  def saveAndNotify(obj: PrismObject): IO[List[String]] = {
    for {
      _ <- storage.save(obj)
      _ <- versionCache.put(s"${obj.id}:${obj.version}", obj)
      notified <- notifySubscribers(obj)
    } yield notified
  }

  /** Notify all subscribers of an object update.
    *
    * @param obj
    *   Updated object
    * @return
    *   List of client IDs that were notified
    */
  def notifySubscribers(obj: PrismObject): IO[List[String]] = {
    for {
      states <- clientStates.get
      notified <- states.toList.flatTraverse { case (clientId, state) =>
        state.getSubscription(obj.id) match {
          case Some(sub) if !sub.temporary =>
            // Update client's version tracking
            state.updateVersion(obj.id, obj.version)
            IO.pure(List(clientId))
          case _ =>
            IO.pure(List.empty[String])
        }
      }
    } yield notified
  }

  /** Get an object from storage, with caching.
    *
    * @param objectId
    *   Object ID
    * @param version
    *   Optional specific version (None = current)
    * @return
    *   The requested object
    */
  def getObject(objectId: String, version: Option[Int] = None): IO[Option[PrismObject]] = {
    version match {
      case Some(v) =>
        val cacheKey = s"$objectId:$v"
        for {
          cached <- versionCache.get(cacheKey)
          result <- cached match {
            case Some(obj) => IO.pure(Some(obj))
            case None =>
              for {
                obj <- storage.getVersion(objectId, v)
                _ <- obj.traverse(o => versionCache.put(cacheKey, o))
              } yield obj
          }
        } yield result

      case None =>
        storage.getCurrent(objectId)
    }
  }

  /** Get a delta between two versions of an object.
    *
    * Uses delta cache for performance.
    *
    * @param objectId
    *   Object ID
    * @param fromVersion
    *   Starting version
    * @param toVersion
    *   Target version
    * @param filterType
    *   Optional filter applied to target object
    * @return
    *   Delta if both versions exist
    */
  def getDelta(
      objectId: String,
      fromVersion: Int,
      toVersion: Int,
      filterType: Option[String] = None
  ): IO[Option[Delta]] = {
    val cacheKey = (objectId, fromVersion, toVersion, filterType)

    for {
      cached <- deltaCache.get(cacheKey)
      result <- cached match {
        case Some(delta) => IO.pure(Some(delta))
        case None =>
          for {
            from <- getObject(objectId, Some(fromVersion))
            to <- getObject(objectId, Some(toVersion))
            delta = (from, to) match {
              case (Some(f), Some(t)) =>
                Some(DeltaComputer.computeDelta(f, t, filterType))
              case _ => None
            }
            _ <- delta.traverse(d => deltaCache.put(cacheKey, d))
          } yield delta
      }
    } yield result
  }

  /** Apply a filter to an object.
    *
    * Uses filter cache for performance.
    *
    * @param obj
    *   Object to filter
    * @param filterType
    *   Optional filter name
    * @param filterParams
    *   Optional filter parameters
    * @return
    *   Filtered object
    */
  def applyFilter(
      obj: PrismObject,
      filterType: Option[String],
      filterParams: Option[ujson.Value] = None
  ): IO[PrismObject] = {
    filterType match {
      case None => IO.pure(obj)
      case Some(ft) =>
        val cacheKey = (obj.id, obj.version, ft, filterParams)

        for {
          cached <- filterCache.get(cacheKey)
          result <- cached match {
            case Some(filtered) => IO.pure(filtered)
            case None =>
              val filter = filterRegistry.get(ft)
              val filtered = filter.apply(obj, filterParams)
              for {
                _ <- if (filter.cacheable) filterCache.put(cacheKey, filtered) else IO.unit
              } yield filtered
          }
        } yield result
    }
  }

  /** Delete an object and notify all subscribers.
    *
    * @param objectId
    *   Object to delete
    * @return
    *   List of client IDs that were notified
    */
  def deleteAndNotify(objectId: String): IO[List[String]] = {
    for {
      _ <- storage.delete(objectId)
      // Remove from caches
      _ <- versionCache.remove(objectId)
      // Notify subscribers - they need to know the object was deleted
      states <- clientStates.get
      notified = states.collect {
        case (clientId, state) if state.getSubscription(objectId).exists(!_.temporary) =>
          state.removeSubscription(objectId)
          clientId
      }.toList
    } yield notified
  }

  /** List all object IDs.
    *
    * @param limit
    *   Maximum number of results
    * @param offset
    *   Starting offset
    * @return
    *   List of object IDs
    */
  def listObjects(limit: Int = 100, offset: Int = 0): IO[List[String]] = {
    storage.listObjects(limit, offset)
  }

  /** Clear temporary subscriptions for a client.
    *
    * This should be called after processing a request to clean up temporary subscriptions.
    *
    * @param clientId
    *   Client identifier
    */
  def clearTemporarySubscriptions(clientId: String): IO[Unit] = {
    for {
      clientState <- getClientState(clientId)
      _ = clientState.clearTemporarySubscriptions()
    } yield ()
  }

  /** Get all active subscriptions for a client.
    *
    * @param clientId
    *   Client identifier
    * @return
    *   List of active (non-temporary) subscriptions
    */
  def getActiveSubscriptions(clientId: String): IO[Iterable[Subscription]] = {
    for {
      clientState <- getClientState(clientId)
    } yield clientState.getActiveSubscriptions
  }

  /** Notify all subscribed clients when an object is updated.
    *
    * This is the key method for real-time synchronization. When an object changes,
    * this method sends delta or fullObject messages to all subscribed clients.
    *
    * @param obj
    *   The updated object
    * @return
    *   Number of clients notified
    */
  def notifyObjectUpdated(obj: PrismObject): IO[Int] = {
    for {
      // Update version cache
      _ <- versionCache.put(s"${obj.id}:${obj.version}", obj)

      // Get all clients and their subscriptions
      states <- clientStates.get

      // Send updates to all subscribed clients
      _ <- states.toList.traverse { case (clientId, state) =>
        state.getSubscription(obj.id) match {
          case Some(subscription) if !subscription.temporary =>
            sendUpdate(clientId, obj, subscription)
          case _ =>
            IO.unit
        }
      }
    } yield states.count { case (_, state) =>
      state.getSubscription(obj.id).exists(!_.temporary)
    }
  }

  /** Send update to a specific client.
    *
    * @param clientId
    *   Client identifier
    * @param obj
    *   Updated object
    * @param subscription
    *   Client's subscription
    */
  private def sendUpdate(clientId: String, obj: PrismObject, subscription: Subscription): IO[Unit] = {
    for {
      // Apply filter
      filtered <- applyFilter(obj, if (subscription.filterType == "default") None else Some(subscription.filterType), subscription.filterParams)

      // Smart sync
      _ <- smartSync(clientId, obj.id, filtered)
    } yield ()
  }

  /** Intelligently sync object with client based on their current state.
    *
    * @param clientId
    *   Client identifier
    * @param objectId
    *   Object ID
    * @param currentObj
    *   Current filtered object
    * @param knownVersion
    *   Version client has (if known)
    */
  private def smartSync(clientId: String, objectId: String, currentObj: PrismObject, knownVersion: Option[Int] = None): IO[Unit] = {
    for {
      clientState <- getClientState(clientId)

      version = knownVersion.orElse(clientState.getVersion(objectId))

      _ <- version match {
        case None =>
          // Client doesn't have object - send full
          sendFullObject(clientId, currentObj)
        case Some(v) if v < currentObj.version =>
          // Client has older version - try delta
          sendDeltaOrFull(clientId, objectId, v, currentObj)
        case _ =>
          // Client has current version, no update needed
          IO.unit
      }

      // Update client's known version
      _ = clientState.updateVersion(objectId, currentObj.version)
    } yield ()
  }

  /** Send delta if efficient, otherwise full object.
    *
    * @param clientId
    *   Client identifier
    * @param objectId
    *   Object ID
    * @param fromVersion
    *   Client's current version
    * @param toObj
    *   Target object
    */
  private def sendDeltaOrFull(clientId: String, objectId: String, fromVersion: Int, toObj: PrismObject): IO[Unit] = {
    for {
      // Try to compute delta
      deltaOpt <- getDelta(objectId, fromVersion, toObj.version)

      _ <- deltaOpt match {
        case Some(delta) if DeltaComputer.isDeltaEfficient(delta, toObj) =>
          // Delta is efficient - send it
          sendDelta(clientId, delta)
        case _ =>
          // Delta too large or unavailable - send full object
          sendFullObject(clientId, toObj)
      }
    } yield ()
  }

  /** Send full object to client.
    *
    * @param clientId
    *   Client identifier
    * @param obj
    *   Object to send
    */
  private def sendFullObject(clientId: String, obj: PrismObject): IO[Unit] = {
    val message = ServerMessage.FullObject(FullObjectMessage(
      id = obj.id,
      version = obj.version,
      data = obj.data,
      filtered = true
    ))
    sendMessage(clientId, message)
  }

  /** Send delta to client.
    *
    * @param clientId
    *   Client identifier
    * @param delta
    *   Delta to send
    */
  private def sendDelta(clientId: String, delta: Delta): IO[Unit] = {
    val message = ServerMessage.Delta(DeltaMessage(
      id = delta.objectId,
      fromVersion = delta.fromVersion,
      toVersion = delta.toVersion,
      patches = delta.patches
    ))
    sendMessage(clientId, message)
  }

  /** Send message to client via callback.
    *
    * @param clientId
    *   Client identifier
    * @param message
    *   Message to send
    */
  private def sendMessage(clientId: String, message: ServerMessage): IO[Unit] = {
    for {
      callbacks <- sendCallbacks.get
      _ <- callbacks.get(clientId) match {
        case Some(callback) => callback(message)
        case None => IO.unit // Client not connected or callback not registered
      }
    } yield ()
  }
}

object ObjectManager {

  /** Create a new ObjectManager.
    *
    * @param storage
    *   Storage adapter
    * @param filterRegistry
    *   Filter registry
    * @param versionCacheSize
    *   Maximum size of version cache (default: 10000)
    * @param deltaCacheSize
    *   Maximum size of delta cache (default: 5000)
    * @param filterCacheSize
    *   Maximum size of filter cache (default: 5000)
    * @return
    *   A new ObjectManager instance
    */
  def create(
      storage: StorageAdapter,
      filterRegistry: FilterRegistry,
      versionCacheSize: Int = 10000,
      deltaCacheSize: Int = 5000,
      filterCacheSize: Int = 5000
  ): IO[ObjectManager] = {
    for {
      clientStates <- Ref.of[IO, Map[String, ClientState]](Map.empty)
      sendCallbacks <- Ref.of[IO, Map[String, ServerMessage => IO[Unit]]](Map.empty)
      versionCache <- LRUCache.create[String, PrismObject](versionCacheSize)
      deltaCache <- LRUCache.create[(String, Int, Int, Option[String]), Delta](deltaCacheSize)
      filterCache <- LRUCache.create[(String, Int, String, Option[ujson.Value]), PrismObject](
        filterCacheSize
      )
    } yield new ObjectManager(storage, filterRegistry, clientStates, sendCallbacks, versionCache, deltaCache, filterCache)
  }
}
